### PR TITLE
CRT S3 Client PutObject

### PR DESCRIPTION
**Pull request by dagnir**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the **CONTRIBUTING** document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have read the **README** document
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [ ] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/s3crt/RequestDataSupplierAdapter.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    private volatile Condition eventAvailable = lock.newCondition();
+    private volatile boolean dataRequested = false;
+
+    public RequestDataSupplierAdapter(Publisher<ByteBuffer> bodyPublisher) {
+        this.bodyPublisher = bodyPublisher;
+        this.bodyPublisher.subscribe(this);
+    }
+
+    @Override
+    public boolean getRequestBytes(ByteBuffer bytes) {
```
Comment (dagnir) [inline services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/s3crt/RequestDataSupplierAdapter.java:31]: I've doing some testing against this updated interface, and CRT doesn't seem to treat the return value the way I'd expect, given this description.

 I would expect the following to work
 
```java
getRequestBytes(ByteBuffer bb) {
    return false;
}
```

which results in `getRequestBytes()` being called continuously, but returning false results in the request completing instead.

Comment (dagnir) [reply services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/s3crt/RequestDataSupplierAdapter.java:31]: Oops, the docs are referring to are here https://github.com/awslabs/aws-crt-java/blob/06593b80e3497a74a27b87a59a9f5c4b7c07cd56/s3-native-client/src/main/java/com/amazonaws/s3/RequestDataSupplier.java#L14-L17

Comment (dagnir) [reply services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/s3crt/RequestDataSupplierAdapter.java]: Discussing with CRT team about what we can do to support the use case above. For now, the implementation fills the buffer before returning

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/s3crt/RequestDataSupplierAdapter.java [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+                eventBuffer.add(new CompleteEvent());
+                eventAvailable.signalAll();
+            } finally {
+                lock.unlock();
+            }
+        }
+    }
+
+    private void waitForSubscription() {
+        while (!hasSubscription())
```
[the remark this answers is not in the document]
Comment (dagnir) [reply services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/s3crt/RequestDataSupplierAdapter.java:188]: What do you think is a sufficient value for this?

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/DefaultS3CrtAsyncClient.java [hunk 3]
```diff
@@ -60,4 +87,60 @@ public void close() {
         s3NativeClient.close();
         configuration.close();
     }
+
+    private static RequestDataSupplier adaptToDataSupplier(AsyncRequestBody requestBody) {
+        return new RequestDataSupplierAdapter(requestBody);
+    }
+
+    private static com.amazonaws.s3.model.PutObjectRequest toCrtPutObjectRequest(PutObjectRequest sdkPutObject) {
```
[the remark this answers is not in the document]
Comment (dagnir) [reply services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/DefaultS3CrtAsyncClient.java:95]: :+1:

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/s3crt/RequestDataSupplierAdapter.java [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+
+            this.subscriber = new SubscriberImpl();
+            bodyPublisher.subscribe(this.subscriber);
+            return true;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private class SubscriberImpl implements Subscriber<ByteBuffer> {
```
[the remark this answers is not in the document]
Comment (dagnir) [reply services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/s3crt/RequestDataSupplierAdapter.java]: Done

File: services/s3/src/it/java/software/amazon/awssdk/services/s3/CrtClientIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    public static void teardown() throws IOException {
+        deleteBucketAndAllContents(TEST_BUCKET);
+        Files.delete(testFile.toPath());
+    }
+
+    @Test
+    public void putObject_objectSentCorrectly() throws IOException, NoSuchAlgorithmException {
+        AsyncRequestBody body = AsyncRequestBody.fromFile(testFile.toPath());
```
[the remark this answers is not in the document]
Comment (dagnir) [reply services/s3/src/it/java/software/amazon/awssdk/services/s3/CrtClientIntegrationTest.java:74]: Done

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/s3crt/RequestDataSupplierAdapter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (zoewangg) [reply services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/s3crt/RequestDataSupplierAdapter.java:220]: Maybe 10 seconds? Just a magic number to not block the thread for too long, which hopefully never happens. WDUT?

[the remark this answers is not in the document]
Comment (dagnir) [reply services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/s3crt/RequestDataSupplierAdapter.java]: Went with 5s before I saw this. Hopefully that's still enough

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/s3crt/RequestDataSupplierAdapter.java [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+            if (eventBuffer.isEmpty() && pending == 0) {
+                pending = DEFAULT_REQUEST_SIZE;
+                subscription.request(pending);
+            }
+
+            Event ev = takeFirstEvent();
+
+            // Discard the event if it's not for the current subscriber
+            if (!ev.subscriber().equals(subscriber)) {
+                continue;
```
Comment (joviegas) [inline services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/s3crt/RequestDataSupplierAdapter.java:92]: Should we log event at DEBUG level for such cases ?

File: services/s3/src/it/java/software/amazon/awssdk/services/s3/CrtClientIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+                .region(DEFAULT_REGION)
+                .build();
+    }
+
+    @After
+    public void methodTeardown() {
+        s3Crt.close();
+    }
+
+    public static void teardown() throws IOException {
```
[the remark this answers is not in the document]
Comment (dagnir) [reply services/s3/src/it/java/software/amazon/awssdk/services/s3/CrtClientIntegrationTest.java:74]: Yep, good catch!

**Review by zoewangg — APPROVED** [review]

**Review by joviegas — APPROVED** [review]

**Comment by dagnir** [comment]
Build failures due to dependency on unreleased version for CRT Java. Changes have been merged to mainline https://github.com/awslabs/aws-crt-java/pull/298

**Pull request merged**